File: go/analyze/rules_locks.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"context"
	"regexp"
	"strings"

	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// Phase A2 — lock-heavy operations.
//
// Each rule here flags a DDL/DML construct that takes an
// AccessExclusive or similarly coarse lock on a relation, and suggests
// the PostgreSQL-native pattern that achieves the same outcome without
// blocking concurrent reads/writes. Keystone targets tier-1 production
// where "five seconds of write downtime" is user-visible; the rules
// are calibrated for that audience.

// -- Rule: no-add-fk-without-not-valid ---------------------------------

// NoAddFKWithoutNotValid flags ALTER TABLE … ADD FOREIGN KEY that does
// not carry NOT VALID. A validating ADD FK scans every row under a
// ShareRowExclusive lock; the NOT VALID form is constant-time, and a
// follow-up VALIDATE CONSTRAINT only takes a RowExclusive lock.
type NoAddFKWithoutNotValid struct{}

func (NoAddFKWithoutNotValid) ID() string { return "no-add-fk-without-not-valid" }
func (NoAddFKWithoutNotValid) Description() string {
	return "ALTER TABLE ADD FOREIGN KEY without NOT VALID scans every row under a lock"
}

var reAddForeignKey = regexp.MustCompile(`(?i)\bALTER\s+TABLE\b[^;]*\bADD\s+(?:CONSTRAINT\s+\w+\s+)?FOREIGN\s+KEY\b`)
var reNotValid = regexp.MustCompile(`(?i)\bNOT\s+VALID\b`)

func (a *NoAddFKWithoutNotValid) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		// File-level scan: NOT VALID may live on a later line of the
		// same ADD FK statement. False positives for multi-statement
		// files with independent ADD FKs are rare and acceptable.
		locs := reAddForeignKey.FindAllStringIndex(f.Body, -1)
		if len(locs) == 0 {
			continue
		}
		if reNotValid.MatchString(f.Body) {
			continue
		}
		for _, loc := range locs {
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(lineOfOffset(f.Body, loc[0])),
				Message: "ALTER TABLE ADD FOREIGN KEY without NOT VALID scans every row " +
					"under ShareRowExclusive. Two-step: ADD CONSTRAINT … FOREIGN KEY … NOT VALID; " +
					"then VALIDATE CONSTRAINT (RowExclusive, concurrent-safe).",
			})
		}
	}
	return out, nil
}

// -- Rule: no-add-check-without-not-valid ------------------------------

// NoAddCheckWithoutNotValid mirrors the FK rule for CHECK constraints.
// Identical lock behaviour; identical two-step workaround.
type NoAddCheckWithoutNotValid struct{}

func (NoAddCheckWithoutNotValid) ID() string { return "no-add-check-without-not-valid" }
func (NoAddCheckWithoutNotValid) Description() string {
	return "ALTER TABLE ADD CHECK without NOT VALID scans every row under a lock"
}

var reAddCheck = regexp.MustCompile(`(?i)\bALTER\s+TABLE\b[^;]*\bADD\s+(?:CONSTRAINT\s+\w+\s+)?CHECK\s*\(`)

func (a *NoAddCheckWithoutNotValid) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		locs := reAddCheck.FindAllStringIndex(f.Body, -1)
		if len(locs) == 0 {
			continue
		}
		if reNotValid.MatchString(f.Body) {
			continue
		}
		for _, loc := range locs {
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(lineOfOffset(f.Body, loc[0])),
				Message: "ALTER TABLE ADD CHECK without NOT VALID validates every row synchronously. " +
					"Two-step: ADD CONSTRAINT … CHECK (…) NOT VALID; then VALIDATE CONSTRAINT.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-unique-constraint-direct ---------------------------------

// NoUniqueConstraintDirect flags ALTER TABLE … ADD CONSTRAINT … UNIQUE
// used *without* the USING INDEX form. Direct ADD UNIQUE builds the
// index under an AccessExclusive lock; the CONCURRENTLY alternative
// (CREATE UNIQUE INDEX CONCURRENTLY + ADD CONSTRAINT … USING INDEX)
// blocks only briefly at the attachment step.
type NoUniqueConstraintDirect struct{}

func (NoUniqueConstraintDirect) ID() string { return "no-unique-constraint-direct" }
func (NoUniqueConstraintDirect) Description() string {
	return "ALTER TABLE ADD CONSTRAINT … UNIQUE without USING INDEX locks writes during build"
}

var reAddUniqueConstraint = regexp.MustCompile(`(?i)\bALTER\s+TABLE\b[^;]*\bADD\s+(?:CONSTRAINT\s+\w+\s+)?UNIQUE\s*\(`)
var reUsingIndex = regexp.MustCompile(`(?i)\bUSING\s+INDEX\b`)

func (a *NoUniqueConstraintDirect) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		locs := reAddUniqueConstraint.FindAllStringIndex(f.Body, -1)
		if len(locs) == 0 {
			continue
		}
		// If the file already uses USING INDEX anywhere, assume the
		// author is following the two-step pattern.
		if reUsingIndex.MatchString(f.Body) {
			continue
		}
		for _, loc := range locs {
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(lineOfOffset(f.Body, loc[0])),
				Message: "Direct ADD UNIQUE builds the backing index under AccessExclusive. " +
					"Two-step: CREATE UNIQUE INDEX CONCURRENTLY idx ON t(col); " +
					"ALTER TABLE t ADD CONSTRAINT c UNIQUE USING INDEX idx;",
			})
		}
	}
	return out, nil
}

// -- Rule: no-set-not-null-direct --------------------------------------

// NoSetNotNullDirect flags ALTER COLUMN … SET NOT NULL without an
// accompanying CHECK NOT VALID helper. The direct form scans every
// row under AccessExclusive; the workaround on PG 12+ is to add a
// CHECK (col IS NOT NULL) NOT VALID, VALIDATE it, then SET NOT NULL
// (which then recognises the proven constraint and is near-instant).
type NoSetNotNullDirect struct{}

func (NoSetNotNullDirect) ID() string { return "no-set-not-null-direct" }
func (NoSetNotNullDirect) Description() string {
	return "ALTER COLUMN SET NOT NULL scans every row under AccessExclusive"
}

var reSetNotNull = regexp.MustCompile(`(?i)\bALTER\s+COLUMN\s+\w+\s+SET\s+NOT\s+NULL\b`)
var reCheckIsNotNull = regexp.MustCompile(`(?i)\bCHECK\s*\(\s*\w+\s+IS\s+NOT\s+NULL\s*\)\s*NOT\s+VALID\b`)

func (a *NoSetNotNullDirect) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		locs := reSetNotNull.FindAllStringIndex(f.Body, -1)
		if len(locs) == 0 {
			continue
		}
		// Author who set up the CHECK NOT VALID helper is already
		// following the recommended pattern — don't double-warn.
		if reCheckIsNotNull.MatchString(f.Body) {
			continue
		}
		for _, loc := range locs {
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(lineOfOffset(f.Body, loc[0])),
				Message: "ALTER COLUMN SET NOT NULL scans every row under AccessExclusive. " +
					"On PG 12+: ADD CHECK (col IS NOT NULL) NOT VALID; VALIDATE CONSTRAINT; " +
					"SET NOT NULL (now instant because the CHECK is proven).",
			})
		}
	}
	return out, nil
}

// -- Rule: no-vacuum-full ----------------------------------------------

// NoVacuumFull flags VACUUM FULL inside a migration. It takes an
// AccessExclusive lock on the relation for the rewrite's duration,
// which is proportional to table size. pg_repack does the same work
// online.
type NoVacuumFull struct{}

func (NoVacuumFull) ID() string { return "no-vacuum-full" }
func (NoVacuumFull) Description() string {
	return "VACUUM FULL rewrites the table under AccessExclusive; use pg_repack instead"
}

var reVacuumFull = regexp.MustCompile(`(?i)\bVACUUM\s+FULL\b`)

func (a *NoVacuumFull) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reVacuumFull, keystonev1alpha1.LintLevelError,
		a.ID(),
		"VACUUM FULL takes AccessExclusive for the rewrite's duration. "+
			"For live tables use pg_repack (online equivalent); for one-offs schedule a maintenance window."), nil
}

// -- Rule: no-cluster --------------------------------------------------

// NoCluster flags the CLUSTER command. Same lock semantics as VACUUM
// FULL; same recommendation.
type NoCluster struct{}

func (NoCluster) ID() string { return "no-cluster" }
func (NoCluster) Description() string {
	return "CLUSTER rewrites the table under AccessExclusive"
}

var reCluster = regexp.MustCompile(`(?i)^\s*CLUSTER\b|\bCLUSTER\s+\w+\s+USING\b`)

func (a *NoCluster) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if reCluster.MatchString(line) {
				out = append(out, Finding{
					Rule:     a.ID(),
					Severity: keystonev1alpha1.LintLevelError,
					File:     f.Name,
					Line:     int32(i + 1),
					Message: "CLUSTER rewrites the table under AccessExclusive. " +
						"Use pg_repack for online equivalent; reserve CLUSTER for maintenance windows.",
				})
			}
		}
	}
	return out, nil
}

// -- Rule: no-reindex-without-concurrently -----------------------------

// NoReindexWithoutConcurrently flags REINDEX TABLE/INDEX without
// CONCURRENTLY (PG 12+). The non-CONCURRENTLY form locks writes.
type NoReindexWithoutConcurrently struct{}

func (NoReindexWithoutConcurrently) ID() string { return "no-reindex-without-concurrently" }
func (NoReindexWithoutConcurrently) Description() string {
	return "REINDEX without CONCURRENTLY blocks writes for the rebuild"
}

var reReindex = regexp.MustCompile(`(?i)\bREINDEX\s+(TABLE|INDEX|SCHEMA|DATABASE|SYSTEM)\b`)
var reReindexConcurrently = regexp.MustCompile(`(?i)\bREINDEX\s+(TABLE|INDEX|SCHEMA|DATABASE|SYSTEM)\s+CONCURRENTLY\b`)

func (a *NoReindexWithoutConcurrently) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if !reReindex.MatchString(line) {
				continue
			}
			if reReindexConcurrently.MatchString(line) {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "REINDEX without CONCURRENTLY takes an exclusive lock for the rebuild's duration. " +
					"Use REINDEX ... CONCURRENTLY on PG 12+ for an online rebuild.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-drop-index-without-concurrently --------------------------

// NoDropIndexWithoutConcurrently flags DROP INDEX without CONCURRENTLY.
// DROP INDEX takes an AccessExclusive lock on the parent table for the
// brief metadata flip; CONCURRENTLY widens that window but avoids
// holding the lock while physical cleanup happens. On hot tables the
// difference is user-visible.
type NoDropIndexWithoutConcurrently struct{}

func (NoDropIndexWithoutConcurrently) ID() string { return "no-drop-index-without-concurrently" }
func (NoDropIndexWithoutConcurrently) Description() string {
	return "DROP INDEX without CONCURRENTLY takes AccessExclusive on the parent table"
}

var reDropIndex = regexp.MustCompile(`(?i)\bDROP\s+INDEX\b`)
var reDropIndexConcurrently = regexp.MustCompile(`(?i)\bDROP\s+INDEX\s+CONCURRENTLY\b`)

func (a *NoDropIndexWithoutConcurrently) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if !reDropIndex.MatchString(line) {
				continue
			}
			if reDropIndexConcurrently.MatchString(line) {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "DROP INDEX without CONCURRENTLY takes AccessExclusive on the parent. " +
					"Use DROP INDEX CONCURRENTLY to drop without blocking concurrent writes.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-lock-table-explicit --------------------------------------

// NoLockTableExplicit flags explicit LOCK TABLE statements. PostgreSQL
// acquires the minimum lock level it needs; explicit LOCK TABLE almost
// always means the author is working around a symptom instead of
// fixing the root cause (concurrent migration, missing index, etc.).
type NoLockTableExplicit struct{}

func (NoLockTableExplicit) ID() string { return "no-lock-table-explicit" }
func (NoLockTableExplicit) Description() string {
	return "LOCK TABLE is rarely the right answer; prefer transactional semantics"
}

var reLockTable = regexp.MustCompile(`(?i)\bLOCK\s+TABLE\b`)

func (a *NoLockTableExplicit) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reLockTable, keystonev1alpha1.LintLevelWarning,
		a.ID(),
		"Explicit LOCK TABLE is almost always a workaround. "+
			"PostgreSQL's transaction machinery acquires the minimum lock; "+
			"if you truly need an exclusive lock, consider restructuring the migration."), nil
}

File: rust/keystone-sdk/src/analyze/rules_locks.rs
// SPDX-License-Identifier: Apache-2.0

//! Phase A2 — lock-heavy operations. Ported from `rules_locks.go`.

use std::sync::LazyLock;

use regex::Regex;

use super::{finding, line_of_offset, statement_matches, Analyzer, Finding, LintLevel, Migration};

macro_rules! re {
    ($name:ident, $pat:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($pat).expect($pat));
    };
}

re!(
    RE_ADD_FOREIGN_KEY,
    r"(?i)\bALTER\s+TABLE\b[^;]*\bADD\s+(?:CONSTRAINT\s+\w+\s+)?FOREIGN\s+KEY\b"
);
re!(RE_NOT_VALID, r"(?i)\bNOT\s+VALID\b");
re!(
    RE_ADD_CHECK,
    r"(?i)\bALTER\s+TABLE\b[^;]*\bADD\s+(?:CONSTRAINT\s+\w+\s+)?CHECK\s*\("
);
re!(
    RE_ADD_UNIQUE_CONSTRAINT,
    r"(?i)\bALTER\s+TABLE\b[^;]*\bADD\s+(?:CONSTRAINT\s+\w+\s+)?UNIQUE\s*\("
);
re!(RE_USING_INDEX, r"(?i)\bUSING\s+INDEX\b");
re!(
    RE_SET_NOT_NULL,
    r"(?i)\bALTER\s+COLUMN\s+\w+\s+SET\s+NOT\s+NULL\b"
);
re!(
    RE_CHECK_IS_NOT_NULL,
    r"(?i)\bCHECK\s*\(\s*\w+\s+IS\s+NOT\s+NULL\s*\)\s*NOT\s+VALID\b"
);
re!(RE_VACUUM_FULL, r"(?i)\bVACUUM\s+FULL\b");
re!(RE_CLUSTER, r"(?i)^\s*CLUSTER\b|\bCLUSTER\s+\w+\s+USING\b");
re!(
    RE_REINDEX,
    r"(?i)\bREINDEX\s+(TABLE|INDEX|SCHEMA|DATABASE|SYSTEM)\b"
);
re!(
    RE_REINDEX_CONCURRENTLY,
    r"(?i)\bREINDEX\s+(TABLE|INDEX|SCHEMA|DATABASE|SYSTEM)\s+CONCURRENTLY\b"
);
re!(RE_DROP_INDEX, r"(?i)\bDROP\s+INDEX\b");
re!(
    RE_DROP_INDEX_CONCURRENTLY,
    r"(?i)\bDROP\s+INDEX\s+CONCURRENTLY\b"
);
re!(RE_LOCK_TABLE, r"(?i)\bLOCK\s+TABLE\b");

/// Shared shape for the "file-level scan + guard regex" rules: emit one
/// finding per match of `re` unless `guard` matches anywhere in the file.
fn file_scan_guarded(
    m: &Migration,
    re: &Regex,
    guard: &Regex,
    rule: &str,
    sev: LintLevel,
    msg: &str,
) -> Vec<Finding> {
    let mut out = Vec::new();
    for f in &m.files {
        let locs: Vec<usize> = re.find_iter(&f.body).map(|mat| mat.start()).collect();
        if locs.is_empty() {
            continue;
        }
        if guard.is_match(&f.body) {
            continue;
        }
        for start in locs {
            out.push(finding(
                rule,
                sev,
                &f.name,
                line_of_offset(&f.body, start),
                msg,
            ));
        }
    }
    out
}

/// no-add-fk-without-not-valid
pub struct NoAddFkWithoutNotValid;
impl Analyzer for NoAddFkWithoutNotValid {
    fn id(&self) -> &'static str {
        "no-add-fk-without-not-valid"
    }
    fn description(&self) -> &'static str {
        "ALTER TABLE ADD FOREIGN KEY without NOT VALID scans every row under a lock"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        file_scan_guarded(
            m,
            &RE_ADD_FOREIGN_KEY,
            &RE_NOT_VALID,
            self.id(),
            LintLevel::Warning,
            "ALTER TABLE ADD FOREIGN KEY without NOT VALID scans every row \
             under ShareRowExclusive. Two-step: ADD CONSTRAINT … FOREIGN KEY … NOT VALID; \
             then VALIDATE CONSTRAINT (RowExclusive, concurrent-safe).",
        )
    }
}

/// no-add-check-without-not-valid
pub struct NoAddCheckWithoutNotValid;
impl Analyzer for NoAddCheckWithoutNotValid {
    fn id(&self) -> &'static str {
        "no-add-check-without-not-valid"
    }
    fn description(&self) -> &'static str {
        "ALTER TABLE ADD CHECK without NOT VALID scans every row under a lock"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        file_scan_guarded(
            m,
            &RE_ADD_CHECK,
            &RE_NOT_VALID,
            self.id(),
            LintLevel::Warning,
            "ALTER TABLE ADD CHECK without NOT VALID validates every row synchronously. \
             Two-step: ADD CONSTRAINT … CHECK (…) NOT VALID; then VALIDATE CONSTRAINT.",
        )
    }
}

/// no-unique-constraint-direct
pub struct NoUniqueConstraintDirect;
impl Analyzer for NoUniqueConstraintDirect {
    fn id(&self) -> &'static str {
        "no-unique-constraint-direct"
    }
    fn description(&self) -> &'static str {
        "ALTER TABLE ADD CONSTRAINT … UNIQUE without USING INDEX locks writes during build"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        file_scan_guarded(
            m,
            &RE_ADD_UNIQUE_CONSTRAINT,
            &RE_USING_INDEX,
            self.id(),
            LintLevel::Warning,
            "Direct ADD UNIQUE builds the backing index under AccessExclusive. \
             Two-step: CREATE UNIQUE INDEX CONCURRENTLY idx ON t(col); \
             ALTER TABLE t ADD CONSTRAINT c UNIQUE USING INDEX idx;",
        )
    }
}

/// no-set-not-null-direct
pub struct NoSetNotNullDirect;
impl Analyzer for NoSetNotNullDirect {
    fn id(&self) -> &'static str {
        "no-set-not-null-direct"
    }
    fn description(&self) -> &'static str {
        "ALTER COLUMN SET NOT NULL scans every row under AccessExclusive"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        file_scan_guarded(
            m,
            &RE_SET_NOT_NULL,
            &RE_CHECK_IS_NOT_NULL,
            self.id(),
            LintLevel::Warning,
            "ALTER COLUMN SET NOT NULL scans every row under AccessExclusive. \
             On PG 12+: ADD CHECK (col IS NOT NULL) NOT VALID; VALIDATE CONSTRAINT; \
             SET NOT NULL (now instant because the CHECK is proven).",
        )
    }
}

/// no-vacuum-full
pub struct NoVacuumFull;
impl Analyzer for NoVacuumFull {
    fn id(&self) -> &'static str {
        "no-vacuum-full"
    }
    fn description(&self) -> &'static str {
        "VACUUM FULL rewrites the table under AccessExclusive; use pg_repack instead"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_VACUUM_FULL,
            LintLevel::Error,
            self.id(),
            "VACUUM FULL takes AccessExclusive for the rewrite's duration. \
             For live tables use pg_repack (online equivalent); for one-offs schedule a maintenance window.",
        )
    }
}

/// no-cluster
pub struct NoCluster;
impl Analyzer for NoCluster {
    fn id(&self) -> &'static str {
        "no-cluster"
    }
    fn description(&self) -> &'static str {
        "CLUSTER rewrites the table under AccessExclusive"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if RE_CLUSTER.is_match(line) {
                    out.push(finding(
                        self.id(),
                        LintLevel::Error,
                        &f.name,
                        (i + 1) as i32,
                        "CLUSTER rewrites the table under AccessExclusive. \
                         Use pg_repack for online equivalent; reserve CLUSTER for maintenance windows.",
                    ));
                }
            }
        }
        out
    }
}

/// no-reindex-without-concurrently
pub struct NoReindexWithoutConcurrently;
impl Analyzer for NoReindexWithoutConcurrently {
    fn id(&self) -> &'static str {
        "no-reindex-without-concurrently"
    }
    fn description(&self) -> &'static str {
        "REINDEX without CONCURRENTLY blocks writes for the rebuild"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_REINDEX.is_match(line) {
                    continue;
                }
                if RE_REINDEX_CONCURRENTLY.is_match(line) {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Warning,
                    &f.name,
                    (i + 1) as i32,
                    "REINDEX without CONCURRENTLY takes an exclusive lock for the rebuild's duration. \
                     Use REINDEX ... CONCURRENTLY on PG 12+ for an online rebuild.",
                ));
            }
        }
        out
    }
}

/// no-drop-index-without-concurrently
pub struct NoDropIndexWithoutConcurrently;
impl Analyzer for NoDropIndexWithoutConcurrently {
    fn id(&self) -> &'static str {
        "no-drop-index-without-concurrently"
    }
    fn description(&self) -> &'static str {
        "DROP INDEX without CONCURRENTLY takes AccessExclusive on the parent table"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_DROP_INDEX.is_match(line) {
                    continue;
                }
                if RE_DROP_INDEX_CONCURRENTLY.is_match(line) {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Warning,
                    &f.name,
                    (i + 1) as i32,
                    "DROP INDEX without CONCURRENTLY takes AccessExclusive on the parent. \
                     Use DROP INDEX CONCURRENTLY to drop without blocking concurrent writes.",
                ));
            }
        }
        out
    }
}

/// no-lock-table-explicit
pub struct NoLockTableExplicit;
impl Analyzer for NoLockTableExplicit {
    fn id(&self) -> &'static str {
        "no-lock-table-explicit"
    }
    fn description(&self) -> &'static str {
        "LOCK TABLE is rarely the right answer; prefer transactional semantics"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_LOCK_TABLE,
            LintLevel::Warning,
            self.id(),
            "Explicit LOCK TABLE is almost always a workaround. \
             PostgreSQL's transaction machinery acquires the minimum lock; \
             if you truly need an exclusive lock, consider restructuring the migration.",
        )
    }
}

#[cfg(test)]
mod tests {
    use super::super::{FileBody, Migration};
    use super::*;

    fn mig(body: &str) -> Migration {
        Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: body.into(),
            }],
            ..Default::default()
        }
    }

    #[test]
    fn add_fk_not_valid_guard() {
        assert_eq!(
            NoAddFkWithoutNotValid
                .check(&mig(
                    "ALTER TABLE t ADD CONSTRAINT fk FOREIGN KEY (a) REFERENCES p(id);"
                ))
                .len(),
            1
        );
        assert_eq!(
            NoAddFkWithoutNotValid
                .check(&mig(
                    "ALTER TABLE t ADD CONSTRAINT fk FOREIGN KEY (a) REFERENCES p(id) NOT VALID;"
                ))
                .len(),
            0
        );
    }

    #[test]
    fn unique_constraint_using_index_guard() {
        assert_eq!(
            NoUniqueConstraintDirect
                .check(&mig("ALTER TABLE t ADD CONSTRAINT u UNIQUE (a);"))
                .len(),
            1
        );
        assert_eq!(
            NoUniqueConstraintDirect
                .check(&mig(
                    "ALTER TABLE t ADD CONSTRAINT u UNIQUE USING INDEX ix;"
                ))
                .len(),
            0
        );
    }

    #[test]
    fn set_not_null_and_vacuum_cluster() {
        assert_eq!(
            NoSetNotNullDirect
                .check(&mig("ALTER TABLE t ALTER COLUMN a SET NOT NULL;"))
                .len(),
            1
        );
        assert_eq!(NoVacuumFull.check(&mig("VACUUM FULL t;")).len(), 1);
        assert_eq!(NoCluster.check(&mig("CLUSTER t USING ix;")).len(), 1);
    }

    #[test]
    fn reindex_and_drop_index_concurrently() {
        assert_eq!(
            NoReindexWithoutConcurrently
                .check(&mig("REINDEX TABLE t;"))
                .len(),
            1
        );
        assert_eq!(
            NoReindexWithoutConcurrently
                .check(&mig("REINDEX TABLE CONCURRENTLY t;"))
                .len(),
            0
        );
        assert_eq!(
            NoDropIndexWithoutConcurrently
                .check(&mig("DROP INDEX ix;"))
                .len(),
            1
        );
        assert_eq!(
            NoDropIndexWithoutConcurrently
                .check(&mig("DROP INDEX CONCURRENTLY ix;"))
                .len(),
            0
        );
    }
}
